xmppserver/src/main/java/org/jivesoftware/openfire/http/HttpSession.java

```diff
@@ -740,39 +754,43 @@ private void processConnection(@Nonnull final HttpConnection connection, @Nonnul
+                // Note that the code leading up to here checks if the Request ID of the new connection 'fits in the window',
                 // which means that for polling sessions, the request ID must have been a sequential one, which in turn should
                 // guarantee that 'a new connection is now available for delivery').
-                assert aConnectionAvailableForDelivery;
+                assert aConnectionAvailableForDelivery; // FIXME: OF-2451: the edge-cases evaluated above make this assertion not necessarily true.
             }
 
-            if (isPollingSession() || (aConnectionAvailableForDelivery && !pendingElements.isEmpty())) {
-                lastActivity = Instant.now();
-                SessionEventDispatcher.dispatchEvent( this, SessionEventDispatcher.EventType.connection_opened, connection, context ); // TODO is this the right place to dispatch this event?
-                deliver(pendingElements);
-                pendingElements.clear();
+            if (isPollingSession() || aConnectionAvailableForDelivery) {
+                // In a thread-safe manner, create a copy of all elements that are currently pending.
+                final List<Deliverable> toProcessElements = drainPendingElements();
+                if (!toProcessElements.isEmpty()) {
+                    lastActivity = Instant.now();
+                    SessionEventDispatcher.dispatchEvent(this, SessionEventDispatcher.EventType.connection_opened, connection, context); // TODO is this the right place to dispatch this event?
+                    deliver(toProcessElements);
+                }
             }
 
             // When a new connection has become available, older connections need to be released (allowing the client to
             // send more data if it needs to).
-            final List<HttpConnection> openConnections = connectionQueue.stream()
-                .filter(c -> !c.isClosed())
-                .filter(c -> c.getRequestId() <= lastSequentialRequestID)
-                .sorted(connectionComparator)
-                .collect(Collectors.toList());
-
-            while( openConnections.size() > hold) {
+            while (!connectionQueue.isEmpty() && connectionQueue.size() > hold) {
                 if (Log.isTraceEnabled()) {
-                    Log.trace("Stream {}: releasing oldest connection (rid {}), as the amount of open connections ({}) is higher than the requested amount to hold ({}).", streamid, rid, openConnections.size(), hold);
+                    Log.trace("Stream {}: releasing oldest connection (rid {}), as the amount of open connections ({}) is higher than the requested amount to hold ({}).", streamid, rid, connectionQueue.size(), hold);
+                }
+                final HttpConnection openConnection = connectionQueue.peek();
+                assert openConnection != null;
+                if (openConnection.getRequestId() > lastSequentialRequestID) {
+                    break; // There's a gap.
```

Comment:
break or continue? Or perhaps a better comment as to what a gap actually means / is interpreted

Reply:
Break. The rationale is provided in comments earlier in the same method, but if you missed that, than others will too. I have modified this comment to explicitly refer to the earlier comments.